# Challenge 2

In this challenge we will continue working with the `Pokemon` dataset. We will attempt solving a slightly more complex problem in which we will practice the iterative data analysis process you leaned in [this video](https://www.youtube.com/watch?v=xOomNicqbkk).

The problem statement is as follows:

**You are at a Pokemon black market planning to buy a Pokemon for battle. All Pokemon are sold at the same price and you can only afford to buy one. You cannot choose which specific Pokemon to buy. However, you can specify the type of the Pokemon - one type that exists in either `Type 1` or `Type 2`. Which type should you choose in order to maximize your chance of receiving a good Pokemon?**

To remind you about the 3 steps of iterative data analysis, they are:

1. Setting Expectations
1. Collecting Information
1. Reacting to Data / Revising Expectations

Following the iterative process, we'll guide you in completing the challenge.


## Problem Solving Iteration 1

In this iteration we'll analyze the problem and identify the breakthrough. The original question statement is kind of vague because we don't know what a *good pokemon* really means as represented in the data. We'll start by understanding the dataset and see if we can find some insights.

In [1]:
# Import libraries
import numpy as np
import pandas as pd

In [2]:
#Here we load the data from the csv file
pokemon = pd.read_csv("Pokemon.csv")

In [3]:
#Here we add the code to display all columns, anticipating to a large volume of columns when using a onehotencoder
pd.set_option('display.max_columns', None) 

From the data it seems whether a pokemon is good depends on its abilities as represented in the fields of `HP`, `Attack`, `Defense`, `Sp. Atk`, `Sp. Def`, `Speed`, and `Total`. We are not sure about `Generation` and `Legendary` because they are not necessarily the decisive factors of the pokemon abilities.

But `HP`, `Attack`, `Defense`, `Sp. Atk`, `Sp. Def`, `Speed`, and `Total` are a lot of fields! If we look at them all at once it's very complicated. This isn't Mission Impossible but it's ideal that we tackle this kind of problem after we learn Machine Learning (which you will do in Module 3). For now, is there a way to consolidate the fields we need to look into?

Fortunately there seems to be a way. It appears the `Total` field is computed based on the other 6 fields. But we need to prove our theory. If we can approve there is a formula to compute `Total` based on the other 6 abilities, we only need to look into `Total`.

We have the following expectation now:

#### The `Total` field is computed based on `HP`, `Attack`, `Defense`, `Sp. Atk`, `Sp. Def`, and `Speed`.

We need to collect the following information:

* **What is the formula to compute `Total`?**
* **Does the formula work for all pokemon?**

In the cell below, make a hypothesis on how `Total` is computed and test your hypothesis.

In [4]:
#Hypothesis: The 'Total' column is the sum of the following columns:HP, Attack, Defense, Sp. Atk, Sp. Def, and Speed

pokemon['Total_2'] = pokemon['HP']+pokemon['Attack']+pokemon['Defense']+pokemon['Sp. Atk']+pokemon['Sp. Def']+pokemon['Speed']

In [5]:
#With the dif_total column we get the difference to validate our hypothesis
pokemon['Dif_Total']= pokemon['Total_2']-pokemon['Total']

In [6]:
pokemon

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  Total_2  Dif_Total  
0         65       65     45           1      False      318          0  
1         80       80     60           1      False      405          0  
2        100      100     80           1      False      525          0  
3        122      120     80           1      False      625          0  
4         60       50     65           1      False      309          0  
..       ...      ...    ...         ...        ...      ...        ...  
795      100      150     50           6       True      600          0  
796      160      110    110           6       True      700          0  
797      150      130     70           6       True      600          0  
798      170      130     80           6       True      680          0  
799      130       90     70           6       True      600          0  

[800 rows x 15 columns]

In [7]:
#If the difference between column total vs total_2 = '0', then our hypothesis is correct 

#With code value counts, we can check if If the difference between both columns = '0'

pokemon['Dif_Total'].value_counts()

Dif_Total
0    800
Name: count, dtype: int64

In [8]:
#HYPOTHESIS IS TRUE
#Given that our hypothesis is true, we can now drop HP, Attack, Defense, Sp. Atk, Sp. Def, and Speed, along with the calculated columns
pokemon = pokemon.drop(['HP','Attack','Defense','Sp. Atk','Sp. Def','Speed','Total_2','Dif_Total'], axis=1)
pokemon

#                   Name   Type 1  Type 2  Total  Generation  Legendary
0      1              Bulbasaur    Grass  Poison    318           1      False
1      2                Ivysaur    Grass  Poison    405           1      False
2      3               Venusaur    Grass  Poison    525           1      False
3      3  VenusaurMega Venusaur    Grass  Poison    625           1      False
4      4             Charmander     Fire     NaN    309           1      False
..   ...                    ...      ...     ...    ...         ...        ...
795  719                Diancie     Rock   Fairy    600           6       True
796  719    DiancieMega Diancie     Rock   Fairy    700           6       True
797  720    HoopaHoopa Confined  Psychic   Ghost    600           6       True
798  720     HoopaHoopa Unbound  Psychic    Dark    680           6       True
799  721              Volcanion     Fire   Water    600           6       True

[800 rows x 7 columns]

In [9]:
#Here we clean the pokemon names just as we did in challenge 1
import re
pokemon['Name'] = pokemon['Name'].apply(lambda x: re.sub(r'(.+)(Mega.+)',r'\2', x))
pokemon

#                 Name   Type 1  Type 2  Total  Generation  Legendary
0      1            Bulbasaur    Grass  Poison    318           1      False
1      2              Ivysaur    Grass  Poison    405           1      False
2      3             Venusaur    Grass  Poison    525           1      False
3      3        Mega Venusaur    Grass  Poison    625           1      False
4      4           Charmander     Fire     NaN    309           1      False
..   ...                  ...      ...     ...    ...         ...        ...
795  719              Diancie     Rock   Fairy    600           6       True
796  719         Mega Diancie     Rock   Fairy    700           6       True
797  720  HoopaHoopa Confined  Psychic   Ghost    600           6       True
798  720   HoopaHoopa Unbound  Psychic    Dark    680           6       True
799  721            Volcanion     Fire   Water    600           6       True

[800 rows x 7 columns]

In [10]:
#Here we replace all 'Nan' values with a string named 'nan', anticipating to have the column name as string when doing the onehotencoding

pokemon.fillna("nan",inplace=True)
pokemon.isna().sum()

#             0
Name          0
Type 1        0
Type 2        0
Total         0
Generation    0
Legendary     0
dtype: int64

## Problem Solving Iteration 2

Now that we have consolidated the abilities fields, we can update the problem statement. The new problem statement is:

### Which pokemon type is most likely to have the highest `Total` value?

In the updated problem statement, we assume there is a certain relationship between the `Total` and the pokemon type. But we have two *type* fields (`Type 1` and `Type 2`) that have string values. In data analysis, string fields have to be transformed to numerical format in order to be analyzed. 

In addition, keep in mind that `Type 1` always has a value but `Type 2` is sometimes empty (having the `NaN` value). Also, the pokemon type we choose may be either in `Type 1` or `Type 2`.

Now our expectation is:

#### `Type 1` and `Type 2` string variables need to be converted to numerical variables in order to identify the relationship between `Total` and the pokemon type.

The information we need to collect is:

#### How to convert two string variables to numerical?

Let's address the first question first. You can use a method called **One Hot Encoding** which is frequently used in machine learning to encode categorical string variables to numerical. The idea is to gather all the possible string values in a categorical field and create a numerical field for each unique string value. Each of those numerical fields uses `1` and `0` to indicate whether the data record has the corresponding categorical value. A detailed explanation of One Hot Encoding can be found in [this article](https://hackernoon.com/what-is-one-hot-encoding-why-and-when-do-you-have-to-use-it-e3c6186d008f). You will formally learn it in Module 3.

For instance, if a pokemon has `Type 1` as `Poison` and `Type 2` as `Fire`, then its `Poison` and `Fire` fields are `1` whereas all other fields are `0`. If a pokemon has `Type 1` as `Water` and `Type 2` as `NaN`, then its `Water` field is `1` whereas all other fields are `0`.

#### In the next cell, use One Hot Encoding to encode `Type 1` and `Type 2`. Use the pokemon type values as the names of the numerical fields you create.

The new numerical variables you create should look like below:

![One Hot Encoding](../images/one-hot-encoding.png)

In [11]:
#With dtypes code we check the type of each column in the dataframe
pokemon.dtypes


#              int64
Name          object
Type 1        object
Type 2        object
Total          int64
Generation     int64
Legendary       bool
dtype: object

In [12]:
#Here we start splitting numerical columns from categorical
numerical = pokemon.select_dtypes(include=[np.number])

#We also drop '#' and 'Generation', to later check correlation with encoded categorical variables
numerical = numerical.drop(['#','Generation'], axis=1)
numerical

Total
0      318
1      405
2      525
3      625
4      309
..     ...
795    600
796    700
797    600
798    680
799    600

[800 rows x 1 columns]

In [13]:
#Here we start splitting categorical columns from numericals. Furthermore we drop 'Name' to prepare them for the OneHotEncoding

categorical = pokemon.select_dtypes(include=[object])
categorical = categorical.drop(['Name'], axis=1)
categorical

Type 1  Type 2
0      Grass  Poison
1      Grass  Poison
2      Grass  Poison
3      Grass  Poison
4       Fire     nan
..       ...     ...
795     Rock   Fairy
796     Rock   Fairy
797  Psychic   Ghost
798  Psychic    Dark
799     Fire   Water

[800 rows x 2 columns]

In [14]:
#Here we encode all categorical features, except for 'Name', into numerical features

from sklearn.preprocessing import OneHotEncoder

encoder = OneHotEncoder().fit(categorical)

#With the following code line we get all the column names in the variable named 'cols'

cols = encoder.get_feature_names_out(input_features=categorical.columns) 

#With the following code we split all column names from '_' and keep only their 1st index to drop Type 1 and 2 out of it
cols = [name.split('_')[1] for name in cols]

categorical_encoded = pd.DataFrame(encoder.transform(categorical).toarray(),columns=cols)

#With groupby, we group those columns where their name are equal, and sum the values to keep those who are =1
categorical_encoded = categorical_encoded.groupby(level=0, axis=1).sum().astype(int)

#Here we reorder the columns to be the same as the expected result, and drop the column named 'nan'
categorical_encoded = categorical_encoded.reindex(columns=['Grass', 'Poison', 'Fire', 'Flying', 'Dragon', 'Water', 'Bug', 'Normal', 'Electric', 'Ground', 'Fairy', 'Fighting','Psychic', 'Rock', 'Steel', 'Ice', 'Ghost', 'Dark'])

#and here we display it to check the result
categorical_encoded



Grass  Poison  Fire  Flying  Dragon  Water  Bug  Normal  Electric  \
0        1       1     0       0       0      0    0       0         0   
1        1       1     0       0       0      0    0       0         0   
2        1       1     0       0       0      0    0       0         0   
3        1       1     0       0       0      0    0       0         0   
4        0       0     1       0       0      0    0       0         0   
..     ...     ...   ...     ...     ...    ...  ...     ...       ...   
795      0       0     0       0       0      0    0       0         0   
796      0       0     0       0       0      0    0       0         0   
797      0       0     0       0       0      0    0       0         0   
798      0       0     0       0       0      0    0       0         0   
799      0       0     1       0       0      1    0       0         0   

     Ground  Fairy  Fighting  Psychic  Rock  Steel  Ice  Ghost  Dark  
0         0      0         0        0     0      0    0      0     0  
1         0      0         0        0     0      0    0      0     0  
2         0      0         0        0     0      0    0      0     0  
3         0      0         0        0     0      0    0      0     0  
4         0      0         0        0     0      0    0      0     0  
..      ...    ...       ...      ...   ...    ...  ...    ...   ...  
795       0      1         0        0     1      0    0      0     0  
796       0      1         0        0     1      0    0      0     0  
797       0      0         0        1     0      0    0      1     0  
798       0      0         0        1     0      0    0      0     1  
799       0      0         0        0     0      0    0      0     0  

[800 rows x 18 columns]

## Problem Solving Iteration 3

Now we have encoded the pokemon types, we will identify the relationship between `Total` and the encoded fields. Our expectation is:

#### There are relationships between `Total` and the encoded pokemon type variables and we need to identify the correlations.

The information we need to collect is:

#### How to identify the relationship between `Total` and the encoded pokemon type fields?

There are multiple ways to answer this question. The easiest way is to use correlation. In the cell below, calculate the correlation of `Total` to each of the encoded fields. Rank the correlations and identify the #1 pokemon type that is most likely to have the highest `Total`.

In [15]:
#To check the correlation, we first concatenate the numerical dataframe with the encoded categorical dataframe
    
consolidated=pd.concat([numerical, categorical_encoded], axis=1)

In [16]:
consolidated

Total  Grass  Poison  Fire  Flying  Dragon  Water  Bug  Normal  Electric  \
0      318      1       1     0       0       0      0    0       0         0   
1      405      1       1     0       0       0      0    0       0         0   
2      525      1       1     0       0       0      0    0       0         0   
3      625      1       1     0       0       0      0    0       0         0   
4      309      0       0     1       0       0      0    0       0         0   
..     ...    ...     ...   ...     ...     ...    ...  ...     ...       ...   
795    600      0       0     0       0       0      0    0       0         0   
796    700      0       0     0       0       0      0    0       0         0   
797    600      0       0     0       0       0      0    0       0         0   
798    680      0       0     0       0       0      0    0       0         0   
799    600      0       0     1       0       0      1    0       0         0   

     Ground  Fairy  Fighting  Psychic  Rock  Steel  Ice  Ghost  Dark  
0         0      0         0        0     0      0    0      0     0  
1         0      0         0        0     0      0    0      0     0  
2         0      0         0        0     0      0    0      0     0  
3         0      0         0        0     0      0    0      0     0  
4         0      0         0        0     0      0    0      0     0  
..      ...    ...       ...      ...   ...    ...  ...    ...   ...  
795       0      1         0        0     1      0    0      0     0  
796       0      1         0        0     1      0    0      0     0  
797       0      0         0        1     0      0    0      1     0  
798       0      0         0        1     0      0    0      0     1  
799       0      0         0        0     0      0    0      0     0  

[800 rows x 19 columns]

In [17]:
#Here we build the correlations matrix between 'Total' and the rest of Encoded columns
correlations_matrix = consolidated.corr()

#With the below code, we sort the correlation and rank each encoded columns vs 'Total'
total_correlation = correlations_matrix['Total'].sort_values(ascending=False)
display(total_correlation)

Total       1.000000
Dragon      0.229705
Psychic     0.124688
Steel       0.109703
Fire        0.078726
Fighting    0.077786
Ice         0.060248
Flying      0.059383
Dark        0.056154
Rock        0.032731
Electric    0.020971
Ground      0.015060
Ghost       0.003641
Water      -0.021665
Fairy      -0.036698
Grass      -0.052592
Poison     -0.090441
Normal     -0.105331
Bug        -0.145781
Name: Total, dtype: float64

In [18]:
#DRAGON is the #1 pokemon type that is most likely to have the highest Total

# Bonus Question

Say now you can choose both `Type 1` and `Type 2` of the pokemon. In order to receive the best pokemon, which types will you choose?

In [21]:
#Identifying all the combinations of Type2 associated to Type1 Dragon:
dragon = pokemon.loc[pokemon['Type 1'] == 'Dragon']
dragon_t2 = dragon['Type 2'].unique()
dragon_t2

array(['nan', 'Flying', 'Fairy', 'Psychic', 'Ground', 'Fire', 'Electric',
       'Ice'], dtype=object)

In [22]:
for type_2_val in dragon_t2:
    print(type_2_val)
    dragon_pok_w_type_2 = pokemon[(pokemon['Type 1'] == 'Dragon') & (pokemon['Type 2'] == type_2_val)]
    display(dragon_pok_w_type_2)

nan


#       Name  Type 1 Type 2  Total  Generation  Legendary
159  147    Dratini  Dragon    nan    300           1      False
160  148  Dragonair  Dragon    nan    420           1      False
406  371      Bagon  Dragon    nan    300           3      False
407  372    Shelgon  Dragon    nan    420           3      False
671  610       Axew  Dragon    nan    320           5      False
672  611    Fraxure  Dragon    nan    410           5      False
673  612    Haxorus  Dragon    nan    540           5      False
682  621  Druddigon  Dragon    nan    485           5      False
774  704      Goomy  Dragon    nan    300           6      False
775  705    Sliggoo  Dragon    nan    452           6      False
776  706     Goodra  Dragon    nan    600           6      False

Flying


#            Name  Type 1  Type 2  Total  Generation  Legendary
161  149       Dragonite  Dragon  Flying    600           1      False
365  334         Altaria  Dragon  Flying    490           3      False
408  373       Salamence  Dragon  Flying    600           3      False
409  373  Mega Salamence  Dragon  Flying    700           3      False
425  384        Rayquaza  Dragon  Flying    680           3       True
426  384   Mega Rayquaza  Dragon  Flying    780           3       True

Fairy


#          Name  Type 1 Type 2  Total  Generation  Legendary
366  334  Mega Altaria  Dragon  Fairy    590           3      False

Psychic


#         Name  Type 1   Type 2  Total  Generation  Legendary
417  380       Latias  Dragon  Psychic    600           3       True
418  380  Mega Latias  Dragon  Psychic    700           3       True
419  381       Latios  Dragon  Psychic    600           3       True
420  381  Mega Latios  Dragon  Psychic    700           3       True

Ground


#              Name  Type 1  Type 2  Total  Generation  Legendary
491  443             Gible  Dragon  Ground    300           4      False
492  444            Gabite  Dragon  Ground    410           4      False
493  445          Garchomp  Dragon  Ground    600           4      False
494  445     Mega Garchomp  Dragon  Ground    700           4      False
794  718  Zygarde50% Forme  Dragon  Ground    600           6       True

Fire


#      Name  Type 1 Type 2  Total  Generation  Legendary
706  643  Reshiram  Dragon   Fire    680           5       True

Electric


#    Name  Type 1    Type 2  Total  Generation  Legendary
707  644  Zekrom  Dragon  Electric    680           5       True

Ice


#                Name  Type 1 Type 2  Total  Generation  Legendary
710  646              Kyurem  Dragon    Ice    660           5       True
711  646  KyuremBlack Kyurem  Dragon    Ice    700           5       True
712  646  KyuremWhite Kyurem  Dragon    Ice    700           5       True

### We would choose Dragon-Fire or Dragon-Ice, since there is a single Pokemon option with a total value of 680, meaning higher chance to buy a high performant Pokemon for battle